### ISSUE TITLE
Access Azure MSSQL with AccessToken from Azure Managed Service Identity (MSI)

### ISSUE THREAD

**Issue by iqmeta** [typeorm/typeorm#4510]
**Issue type:**

[x] question
[ ] bug report
([x] feature request)
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[x] `mssql`  (AZURE)
[ ] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[x] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

Hi,
with the .net Framework 4.6 it is possible to Access / Connect to
an MSSQL with a Token...

https://docs.microsoft.com/en-gb/azure/active-directory/managed-identities-azure-resources/tutorial-windows-vm-access-sql

like:
```
//
// Open a connection to the SQL server using the access token.
//
if (accessToken != null) {
    string connectionString = "Data Source=<AZURE-SQL-SERVERNAME>; Initial Catalog=<DATABASE>;";
    SqlConnection conn = new SqlConnection(connectionString);
    conn.AccessToken = accessToken;
    conn.Open();
}

```

Any idea or chance to do this with typeorm as well? 🙈

Cheers Otto.

**Comment by iqmeta** [comment]
Related, Sub-Dependency MSSQL Package:
Support for AccessToken?
https://github.com/tediousjs/tedious/issues/745

**Comment by iqmeta** [comment]
Hi, beta.1 has been released for this issue:
https://github.com/tediousjs/node-mssql/releases/tag/v6.0.0-beta.1

worth a try 🤓

**Comment by funkydev** [comment]
Is there any update? Our team considers to change ORM from Sequelize to TypeORM in our application, and we can't do that, because there is no Access Token Authentication. 

Azure provides a pretty cool library for receiving the access token using different ways of Azure Active Directory Authentication. You can read more about it here: [@azure/identity](https://www.npmjs.com/package/@azure/identity)

I can help if you have any questions. 👀

**Comment by revitalist** [comment]
Any update?  Would be great if TypeORM supported token-based authentication for SQL database in Azure.

**Comment by hnaidin** [comment]
+1 for token-based authentication

**Comment by panbhatt** [comment]
@iqmeta  : I am able to connect to Azure SQL via an MSI token generated from Azure tool. Let me know, if i can help. I am using TSED as a framework and typeorm.

**Comment by funkydev** [comment]
I see that TypeORM uses [`mssql`](https://github.com/tediousjs/node-mssql) package as a driver, and `mssql` uses [`tedious`](https://github.com/tediousjs/tedious) package which supports Managed Service Identities.

I opened PR (https://github.com/tediousjs/node-mssql/pull/1208) for the `mssql` package to support Access Token Authentication for database connections. That should allow to receive and pass access token using [`@azure/identity`](https://www.npmjs.com/package/@azure/identity) package.

Depends on the future of the `mssql` package, it would be nice to refactor interfaces of SQLServer Credentials in TypeORM to support all authentication methods supported by `tedious`. 

I'll keep you up to date about any changes.

**Comment by funkydev** [comment]
Here is copy/paste my proposal from the PR.

### Overview
When I was modifying `mmsql` package, I noticed that it accepts `authentication` parameters in the config object. See discussion under PR https://github.com/tediousjs/node-mssql/pull/1208. It passes authentication configuration directly to the `tedious` library, which supports the following authentication types.

- `DefaultAuthentication`
- `NtlmAuthentication`
- `AzureActiveDirectoryAccessTokenAuthentication`
- `AzureActiveDirectoryMsiAppServiceAuthentication`
- `AzureActiveDirectoryMsiVmAuthentication`
- `AzureActiveDirectoryPasswordAuthentication`
- `AzureActiveDirectoryServicePrincipalSecret`

I copied authentication options interfaces directly from the `tedious` repository and modified TypeORM code to support passing authentication options for the sqlserver driver.

### What does it change?
It changes a way of passing credentials to the SQLServer database. I modified current interfaces to deprecate `domain` property directly in the configuration model. Instead of this I proposed to pass `authentication` property which is compatible with the `mssql` and `tedious` one. 

I'd recommend to remove `options.domain` in future, to give developers time for any adjustments.  Then instead of passing: 
```javascript
createConnection({
  type: "mssql",
  database: "db-name",
  host: "db-server",
  username: "username",
  password: "password",
  domain: "domain"
}).then(connection => {
  // here you can start to work with your entities
}).catch(error => console.log(error));
```
developers should pass:
```javascript
createConnection({
  type: "mssql",
  database: "db-name",
  host: "db-server",
  authentication: {
    type: "ntlm",
    options: {
        userName: "username",
        password: "password",
        domain: "domain"
    }
  }
}).then(connection => {
  // here you can start to work with your entities
}).catch(error => console.log(error));
```

I created the PR for updating `mssql` documentation for `tedious` driver: https://github.com/tediousjs/node-mssql/pull/1209. At the moment you can read the documentation [here](https://github.com/funkydev/node-mssql/tree/feature/authentication-docs#tedious).

### Solves
- https://github.com/typeorm/typeorm/issues/4510 
- https://github.com/typeorm/typeorm/issues/7340

### Usage

#### Default authentication type
```javascript
createConnection({
  type: "mssql",
  database: "db-name",
  host: "db-server",
  authentication: {
    type: "default",
    options: {
      userName: "username",
      password: "password"
    }
  }
}).then(connection => {
  // here you can start to work with your entities
}).catch(error => console.log(error));
```


#### AAD Access Token
```javascript
createConnection({
  type: "mssql",
  database: "db-name",
  host: "db-server",
  authentication: {
    type: "azure-active-directory-access-token",
    options: {
      token: "token"
    }
  }
}).then(connection => {
  // here you can start to work with your entities
}).catch(error => console.log(error));
```


#### AAD Service Principal (secret)
```javascript
createConnection({
  type: "mssql",
  database: "db-name",
  host: "db-server",
  authentication: {
    type: "azure-active-directory-service-principal-secret",
    options: {
      tenantId: 'tenant-id',
      clientId: 'client-id',
      clientSecret: 'client-secret'
    }
  }
}).then(connection => {
  // here you can start to work with your entities
}).catch(error => console.log(error));
```

#### AAD MSI App Service Authentication
```javascript
createConnection({
  type: "mssql",
  database: "db-name",
  host: "db-server",
  authentication: {
    type: "azure-active-directory-msi-app-service",
    options: {
      clientId: "optional clientId (it should be taken from app service environment)",
      msiEndpoint: "optional msiEndpoint (it should be taken from app service environment)",
      msiSecret: "optional msiSecret (it should be taken from app service environment)"
    }
  }
}).then(connection => {
  // here you can start to work with your entities
}).catch(error => console.log(error));
```

**Comment by funkydev** [comment]
@imnotjames You can decide what should happen with that.

**Comment by AlexMesser** [comment]
implemented in #7477
